poanetwork/blockscout
# Get transaction status and error from last internal transaction

Fixes #673

## Test Report

1. Against Ethereum Classic (private nodes)
   1. Run until a transactions with error is found: `transaction = Explorer.Repo.one(from t in Explorer.Chain.Transaction, where: not is_nil(t.error), limit: 1)`
       ```
      iex(6)> transaction.block_number
      6557515
      iex(7)> transaction.hash |> to_string
      "0x2296b765ab869497a8b8d386a8e81db792269bcd12af4d2d1c7a83e5c064a6bc"
      ```
   2. Confirm error is a realistic reason: `transaction.error`
      ```
      iex(8)> transaction.error
      "Bad instruction"
      ```
2. Against POA Core
   1. Run until a transactions with error is found: `transaction = Explorer.Repo.one(from t in Explorer.Chain.Transaction, where: not is_nil(t.error), limit: 1)`
      ```
      iex(7)> transaction.block_number
      4501839
      iex(8)> transaction.hash |> to_string()
      "0xeb90950b417b16366955122e4535ecdadbe16ad65830e6cd99c2349de69cf30e"
      ```
   2. Confirm error is a realistic reason: `transaction.error`:
      ```
      iex(9)> transaction.error
      "Reverted"
      ``` 
    
## Changelog

### Enhancements
* Denormalize last (`max(index)`) internal transaction's `error` to new `error` column on the transaction.  This way, we can display the exact error message and not a simple "Failure".
  * `Explorer.Chain.transaction_to_status` uses `error` to give actual `reason` when a transaction fails:
    * `block_hash: nil, status: nil` -> `:pending` [Unchanged]
    * ` status: nil` -> `:awaiting_internal_transactions` pre-Byzantium/ETC [Added]
    * `status: :ok` -> `:success`[Unchanged]
    * `status: :error, internal_transactions_indexed_at: nil, error: nil` -> `{:error, :awaiting_internal_transactions}` post-Byzantium [Added]
    * `status: :error, error: reason` -> `{:error, reason}` All chains and hard-forks [Added]

### Bug Fixes
* Don't infer status based on `gas_limit == gas_used`, it is an imperfect heuristic, @acravenho has made transactions on ETC where `gas_limit == gas_used` has been a success even though the heuristic would have said it was a failure.
  
### Incompatible Changes
* There is no specific "Out of Gas" status from `BlockScoutWeb.TransactionView.format_status`, but `"Error: Out of Gas"` may be returned when the last internal transaction's error is `"Out of Gas"`, but that is directly from the node's reported format and not normalized as it was before.
* `BlockScoutWeb.Transactionview.status` has been changed to `BlockScoutWeb.TransactionView.status_class` and it now returns the full class name as the old way could not handle tuples returned by `Explorer.Chain.transaction_to_status`.
    * `:pending` -> `"Pending"` [Unchanged]
    * `:awaiting_internal_transactions` -> `"(Awaiting internal transactions for status)"` pre-Byzantium/ETC [Added]
    * `:success` - > `"Success"` [Unchanged]
    * `{:error, :awaiting_internal_transactions}` -> `"Error: (Awaiting internal transactions for reason)"`  post-Byzantium [Added]
    * `"Out of Gas"` [Removed]
    * `"Failure"` [Removed]
    * `{:error, reason}` -> `"Error: %{reason}"` All chains and hard-forks [Added]

## Upgrading

This is a change to the database schema to further enforce when `internal_transactions_indexed_at` and `status` can be set and it adds the `error` field:

1. `mix do ecto.drop, ecto.create, ecto.migrate`
2. `MIX_ENV=test mix do ecto.drop, ecto.create, ecto.migrate`
3. Reset logs: `rm -rf logs`
4. Re-index: `iex -S mix`
